### ISSUE TITLE
SystemError: Bad call flags in PyCFunction_Call. METH_OLDARGS is no longer supported!

### ISSUE THREAD

**Issue by picca** [PierreRaybaut/PythonQwt#25]
While testing our application with the new python-qwt I got this error message under python3.

It works with python2

the version are

picca@sixs7:~/sida/sida/sixs$ dpkg -l | grep qwt
ii  python-guiqwt                         3.0.0~b5-1~exp1                      amd64        efficient 2D data-plotting library - Python 2
ii  python-guiqwt-doc                     3.0.0~b5-1~exp1                      all          efficient 2D data-plotting library - Documentation
ii  python-qwt                            0.2.1-1                              all          Pure Python implementation of Qwt - Python 2
ii  python-qwt-doc                        0.2.1-1                              all          Pure Python implementation of Qwt - Documentation
ii  python3-guiqwt                        3.0.0~b5-1~exp1                      amd64        efficient 2D data-plotting library - Python 3
ii  python3-qwt                           0.2.1-1                              all          Pure Python implementation of Qwt - Python 3

Traceback (most recent call last):
  File "fitnxs.py", line 1372, in <module>
    main()
  File "fitnxs.py", line 1367, in main
    ui.setupUi(MainWindow)
  File "fitnxs.py", line 1002, in setupUi
    self.makewindow()
  File "fitnxs.py", line 1252, in makewindow
    ylabel="ylabel"))
  File "/usr/lib/python3/dist-packages/guiqwt/plot.py", line 866, in __init__
    panels=panels)
  File "/usr/lib/python3/dist-packages/guiqwt/plot.py", line 781, in __init__
    self.create_plot(options)
  File "/usr/lib/python3/dist-packages/guiqwt/plot.py", line 832, in create_plot
    self.plot_widget = BaseCurveWidget(self, **options)
  File "/usr/lib/python3/dist-packages/guiqwt/plot.py", line 727, in __init__
    section=section, gridparam=gridparam)
  File "/usr/lib/python3/dist-packages/guiqwt/curve.py", line 1339, in __init__
    super(CurvePlot, self).__init__(parent, section)
  File "/usr/lib/python3/dist-packages/guiqwt/baseplot.py", line 163, in __init__
    super(BasePlot, self).__init__(parent)
  File "/usr/lib/python3/dist-packages/qwt/plot.py", line 318, in __init__
    self.initAxesData()
  File "/usr/lib/python3/dist-packages/qwt/plot.py", line 378, in initAxesData
    d.scaleWidget.setTransformation(d.scaleEngine.transformation())
  File "/usr/lib/python3/dist-packages/qwt/scale_widget.py", line 711, in setTransformation
    self.layoutScale()
  File "/usr/lib/python3/dist-packages/qwt/scale_widget.py", line 457, in layoutScale
    bd0, bd1 = self.getBorderDistHint()
  File "/usr/lib/python3/dist-packages/qwt/scale_widget.py", line 643, in getBorderDistHint
    start, end = self.__data.scaleDraw.getBorderDistHint(self.font())
  File "/usr/lib/python3/dist-packages/qwt/scale_draw.py", line 577, in getBorderDistHint
    s = -self.labelRect(font, minTick).top()
  File "/usr/lib/python3/dist-packages/qwt/scale_draw.py", line 1037, in labelRect
    lbl = self.tickLabel(font, value)
  File "/usr/lib/python3/dist-packages/qwt/scale_draw.py", line 420, in tickLabel
    lbl.textSize(font)
  File "/usr/lib/python3/dist-packages/qwt/text.py", line 529, in textSize
    ) = self.__data.textEngine.textMargins(font)
  File "/usr/lib/python3/dist-packages/qwt/text_engine.py", line 264, in textMargins
    top = fm.ascent() - self.effectiveAscent(font)
  File "/usr/lib/python3/dist-packages/qwt/text_engine.py", line 223, in effectiveAscent
    return ASCENTCACHE.setdefault(fontKey, self.findAscent(font))
  File "/usr/lib/python3/dist-packages/qwt/text_engine.py", line 243, in findAscent
    line = img.scanLine(row).asstring(linebytes)
SystemError: Bad call flags in PyCFunction_Call. METH_OLDARGS is no longer supported!

**Comment by PierreRaybaut** [comment]
I have tested this part of the code with the following script with Python 2.7 and Python 3.4 and it works:
(what version of Python 3 are you using?)
```python
import struct
from qwt.qt.QtGui import (QApplication, QColor, QFont, QPixmap, QFontMetrics,
                          QPainter)
from qwt.qt.QtCore import Qt


def findAscent(font):
    
    dummy = "E"
    white = QColor(Qt.white)
    
    fm = QFontMetrics(font)
    pm = QPixmap(fm.width(dummy), fm.height())
    pm.fill(white)
    
    p = QPainter(pm)
    p.setFont(font)
    p.drawText(0, 0, pm.width(), pm.height(), 0, dummy)
    p.end()
    
    img = pm.toImage()
    
    w = pm.width()
    linebytes = w*4
    for row in range(img.height()):
        line = img.scanLine(row).asstring(linebytes)
        for col in range(w):
            color = struct.unpack('I', line[col*4:(col+1)*4])[0]
            if color != white.rgb():
                return fm.ascent()-row+1
                
    return fm.ascent()

if __name__ == '__main__':
    app = QApplication([])
    font = QFont('Arial')
    print(findAscent(font))
```

**Comment by PierreRaybaut** [comment]
Tested on Python 3.3 on Windows 10. It works too.

**Comment by picca** [comment]
com-sixs@sixs7:~/picca/bugs$ python3.4 test.py
Traceback (most recent call last):
  File "test.py", line 30, in <module>
    print(findAscent(font))
  File "test.py", line 20, in findAscent
    line = img.scanLine(row).asstring(linebytes)
SystemError: Bad call flags in PyCFunction_Call. METH_OLDARGS is no longer supported!

**Comment by picca** [comment]
I split the line and it failes with the asstring method

Traceback (most recent call last):
  File "test.py", line 32, in <module>
    print(findAscent(font))
  File "test.py", line 22, in findAscent
    line = sline.asstring(linebytes)
SystemError: Bad call flags in PyCFunction_Call. METH_OLDARGS is no longer supported!

**Comment by picca** [comment]
https://bugs.python.org/issue15657

I think that there is a bad interaction with sip

**Comment by picca** [comment]
https://riverbankcomputing.com/pipermail/pyqt/2014-December/035201.html

**Comment by PierreRaybaut** [comment]
Ok, I think you found the source of the problem.

Given the fact that Phil Thompson wrote [here](https://riverbankcomputing.com/pipermail/pyqt/2014-December/035203.html) that the bug was fixed in latest sip snapshot on december 2014, it should be fixed with PyQt >= 4.11.4 and PyQt >= 5.4.

So the question is: what version of PyQt are you using? and would it be possible for you to upgrade your PyQt installation?

**Comment by picca** [comment]
ii  python-sip                            4.16.4+dfsg-1                        amd64        Python/C++ bindings generator runtime library
ii  python3-sip                           4.16.4+dfsg-1                        amd64        Python 3/C++ bindings generator runtime library
picca@sixs7:~/Debian/guiqwt$ dpkg -l | grep python-qt4
ii  python-qt4                            4.11.2+dfsg-1                        amd64        Python bindings for Qt4
ii  python-qt4-gl                         4.11.2+dfsg-1                        amd64        Python bindings for Qt4's OpenGL module

The current Debian jessie is affected by this bug :((.  It means that for now it is not possible to backport the python3 version of python-qwt. Not a big deal, but it is nice to have the info somewhere :)

thanks
